Make it easier for companies to fund open source

whit537 opened the pull request:
So the basic idea evolving on this PR is to walk the entire dependency graph of a company's source code, and provide a way to give money to the entire bundle at once. We need the projects to opt in first, of course; our first attempt at streamlining that is in #4148. The current plan once #4148 is in is to [use](https://github.com/gratipay/gratipay.com/pull/4135#issuecomment-254896049) the Libraries.io API to get the dependency graph based on a company's uploaded `package.json` file.

whit537 commented:
> Closes #2695.

Actually, nevermind that. I was gonna say that we should also have a nice onboarding page for companies ("Step 1, Step 2"), but since [we only have one company giving right now](https://github.com/gratipay/inside.gratipay.com/issues/844#issuecomment-251934503), I think we should just make the homepage the onboarding page. 😶

whit537 commented:
![screen shot 2016-10-07 at 10 34 33 pm](https://cloud.githubusercontent.com/assets/134455/19209698/5ac951e6-8cde-11e6-9580-0da44f68da91.png)

whit537 commented:
https://github.com/gratipay/gratipay.com/pull/4135/commits/92d411f102aadcf348ff271c3681d16ea64ee786

![screen shot 2016-10-07 at 11 00 32 pm](https://cloud.githubusercontent.com/assets/134455/19209880/e210fda4-8ce1-11e6-9d3e-a6dadb940c12.png)

mattbk commented:
Blog post about this when it goes live? Part of Step 6 (Tell the world!) should be "tell the projects you use to get on Gratipay."

whit537 commented:
Step 4 should be "link your GitHub and we will crawl your private repos to find out what packages you depend on."

Then the company sets a dollar amount and we split it equally between all of the packages in the dependency tree that have been claimed. 

We should crawl all the packages repos for metadata and make it available at `/on/{pypi,npm,cpan,cocoa-pods,etc}/package`.

whit537 commented:
Actually, that should be step 1. That's the interesting thing, the hook. I'm a CTO at a small to midling software company, it's [interesting](https://github.com/gratipay/inside.gratipay.com/issues/844#issuecomment-253926018) to me to know what my open source inventory is and especially what licenses are used. On the result page we prompt them to take the next step and fund the projects listed in the report.

We call it a "bundle" or something: the set of open source packages you depend on it. Each week we recrawl to generate a fresh list, and we add a step to payday to distribute an amount equally across teams (projects) in a bundle.

Then we need some way for package owners to claim their package on Gratipay, i.e., make it a team. I guess for demo purposes we can do that manually in the review ticket. The "Claim" button on the `/on/pypi/aspen` page can simply `POST` to `/new` with the form partially filled out, and they can finish applying there. They're into the existing team flow at that point.

whit537 commented:
Restarting this PR with a different tack. Previous head was 92d411f102aadcf348ff271c3681d16ea64ee786.

whit537 commented:
![screen shot 2016-10-16 at 10 09 21 pm](https://cloud.githubusercontent.com/assets/134455/19423325/3b2a8042-93ed-11e6-93fc-7a08f5a51fde.png)

whit537 commented:
![screen shot 2016-10-16 at 10 10 30 pm](https://cloud.githubusercontent.com/assets/134455/19423337/5beaea6a-93ed-11e6-9c67-c1e7cab3c9ce.png)

whit537 commented:
I constrained the "1.0 payouts" notice to just pages under ~username (there's no other way to access the page, so we need to keep _something_). The rest of the time there is now a "Gratipay is a great way to [fund open source]()" message that points to the dedicated page.

![screen shot 2016-10-16 at 10 10 59 pm](https://cloud.githubusercontent.com/assets/134455/19423354/8ef2f68c-93ed-11e6-8fef-6e8fe0d96da0.png)

whit537 commented:
https://martin-thoma.com/analyzing-pypi-metadata/
https://martin-thoma.com/analyzing-pypi-metadata-2/

whit537 commented:
Version 0 should probably just take a requirements.txt in a textarea.

whit537 commented:
I really want to demo this in my lightning talk at [ATO](https://github.com/gratipay/inside.gratipay.com/issues/757). I need to get slides tomorrow! 😮

whit537 commented:
![screen shot 2016-10-16 at 10 31 34 pm](https://cloud.githubusercontent.com/assets/134455/19423633/51ebff06-93f0-11e6-819f-36edce996196.png)

whit537 commented:
We could probably replace the welcome notice about the number of teams with a link to Fund Open Source.

whit537 commented:
I'm looking at the next part, where we take in `requirements.txt` and we want to spit out a list of all of the dependencies (not just the first-level ones, _all_ of them), as well as their PyPI names and licenses. I had thought maybe we could just use `pip` programmatically to parse the requirements, but we need to actually install everything in order to see what's installed. That means we need some sort of sandboxing, because installing packages involves executing arbitrary code. Hmm ...

whit537 commented:
I'm looking at building a micro-service like we have for [processing images](http://gip.rocks/). I'm looking at using [Docker](https://www.docker.com/) for the sandboxing, hosted on a Digital Ocean droplet.

whit537 commented:
Blam!

```
root@gdr:~# docker run gdr ./licenselife.py
Greetings, program!
root@gdr:~#
```

Docker is nice. :-)

whit537 commented:
👍

```
root@gdr:~# docker run gdr ./gdr.py
click                6.6                  UNKNOWN             
Flask                0.11.1               BSD                 
itsdangerous         0.24                 UNKNOWN             
Jinja2               2.8                  BSD                 
MarkupSafe           0.23                 BSD                 
Werkzeug             0.11.11              BSD                 
root@gdr:~# docker run gdr cat requirements.txt
Flask
root@gdr:~#
```

```python
#!/usr/bin/env python
from __future__ import absolute_import, division, print_function, unicode_literals

import json
import subprocess

out = lambda cmd: subprocess.check_output(cmd.split()).decode('ascii')


def javascript():
    things = out('find node_modules -name package.json')
    for thing in things.splitlines():
        metadata = json.load(open(thing))
        package = metadata.get('name', 'n/a')
        version = metadata.get('version', 'n/a')
        license = metadata.get('license', 'n/a')
        if type(license) is dict:
            license = license.get('type', 'n/a')
        print('{:<30} {:<20} {:<20}'.format(package, version, license))


def python():
    out("rm -rf env")
    out("virtualenv env")
    out("env/bin/pip install -r requirements.txt")
    things = out("env/bin/pip freeze")

    for line in things.splitlines():
        package, version = line.split('==')
        path = 'env/lib/python2.7/site-packages/{}-{}.dist-info/METADATA'.format(package, version)
        license = 'n/a'
        for line in open(path):
            if line.startswith('License:'):
                license = line.split()[1]
        print('{:<20} {:<20} {:<20}'.format(package, version, license))


python()
```

whit537 commented:
Can those names be trusted as the PyPI names?

whit537 commented:
The repo for this will be a mashup of gip.rocks (microservice) and grtp.co (Digital Ocean droplet).

whit537 commented:
I bought `gdr.rocks` ($19.90) for this.

whit537 commented:
Graitpay Dependency Resolver! 💃

whit537 commented:
Alright, not gonna mess with docker-in-docker for this. Looks like the path of least resistance is going to be using DO's Docker [image](https://www.digitalocean.com/community/tutorials/how-to-use-the-digitalocean-docker-application).

whit537 commented:
I hope this works. 😳

whit537 commented:
I feel like this is a really important moment for Gratipay. This might—_might_—work!

whit537 commented:
Repo is up: https://github.com/gratipay/gdr.rocks.

Will circle back here after https://github.com/gratipay/gdr.rocks/issues/1.

whit537 commented:
😍 

![screen shot 2016-10-17 at 8 26 26 pm](https://cloud.githubusercontent.com/assets/134455/19460264/13951fea-94a8-11e6-943b-e5d25baab703.png)

whit537 commented:
Hmmm ... I'm working through the Read the Docs example. I had to install `git` for the Git links, and now I'm installing `libxml2-dev` and `libxslt1-dev` for `lxml`. Reminds me of the cryptography [shenanigans](https://github.com/heroku/heroku-buildpack-python/blob/ba7a9cd565aa1b71ffc5274f37dad5e7cb4894e6/bin/steps/cryptography#L13) in heroku-buildpack-python (cf. #4038). I feel like I'm teetering over the edge of a [Travis-CI rabbit hole](https://docs.travis-ci.com/user/customizing-the-build#Installing-Packages-Using-apt).

whit537 commented:
> Further Ideas
>
**Build a dependency graph**: Some of the code was already written. However, one has to download about 25 GB of data, extract it and run over those files. This is quite a bit of work.

https://martin-thoma.com/analyzing-pypi-metadata/#further-ideas

whit537 commented:
https://pypi.python.org/pypi/pipdeptree/0.3
&rarr; https://github.com/pypa/pip/issues/988

https://ogirardot.wordpress.com/2013/01/05/state-of-the-pythonpypi-dependency-graph/
&rarr; https://github.com/ogirardot/meta-deps/blob/master/pypi-metadata.py
&rarr; &rarr; http://ogirardot.github.io/meta-deps/ (also Maven)

![pypi-deps](https://cloud.githubusercontent.com/assets/134455/19461270/3638322c-94b1-11e6-8cc4-3378c8918197.png)

whit537 commented:
http://kgullikson88.github.io/blog/pypi-analysis.html
https://github.com/sbenthall/pypi-dependency-graph &larr; fairly recent data; combine w/ rss/changelog [API](https://wiki.python.org/moin/PyPIXmlRpc)?

whit537 commented:
Hrm. I think ` && apt-get build-dep -y lxml` fixed that problem, but the RtD test still fails with a bunch of warnings etc. that I thought weren't important. I guess they are! And my disk is full. 😞

whit537 commented:
> I feel like I'm teetering on the edge of a [Travis-CI rabbit hole](https://docs.travis-ci.com/user/customizing-the-build#Installing-Packages-Using-apt).

Also: streaming output (also, RtD).

whit537 commented:
But really that's optimization in the wrong direction. Better to load the dependency graph into a database for fast queries, rather than streaming slow CPU-churners.

whit537 commented:
All that extra crap sure does load up the image. :-/

```
[gdr]$ docker images
REPOSITORY          TAG                 IMAGE ID            CREATED             SIZE
gdr                 latest              94e6e774afa0        12 minutes ago      850.3 MB
ubuntu              latest              f753707788c5        4 days ago          127.2 MB
[gdr]$
```

whit537 commented:
Another reason to use a graph instead of `pip install` is that we want to look at `setup_requires` and `test_requires` in addition to `install_requires`. It strikes me that we could run into some "Google changed their algorithm and seriously moved my cheese!" scenarios if/when we modify the behavior of bundles. Like, what about OS-level dependencies? All of these apt-gets should be included somehow in the distribution. That's to say nothing of developer tools that don't touch production requirements (`test_requires` only goes so far—what about Vim and Emacs?).

whit537 commented:
Also bear in mind that `requirements.txt` is [different](https://packaging.python.org/requirements/) than the arguments to `setup`. I think starting with `requirements.txt` is the right way to go for us here, since the point is to inventory what end-of-line company projects are using from the ecosystem.

whit537 commented:
And then there's [`extras_require`](http://setuptools.readthedocs.io/en/latest/setuptools.html#declaring-extras-optional-features-with-their-own-dependencies) ... I guess this is the [doc to read](http://setuptools.readthedocs.io/en/latest/setuptools.html#declaring-dependencies).

whit537 commented:
Drat. Working locally, but getting a 500 from production. Might actually be a memory error? Skunked for tonight ... :-/

![screen shot 2016-10-18 at 12 10 00 am](https://cloud.githubusercontent.com/assets/134455/19464279/f06f4952-94c8-11e6-9f3b-3d6e93d33359.png)

![screen shot 2016-10-18 at 12 22 10 am](https://cloud.githubusercontent.com/assets/134455/19464280/f0716ac0-94c8-11e6-8c9d-7e086664d984.png)

whit537 commented:
I had though of [Sentry](https://github.com/getsentry/sentry) as another example, but they don't really use requirements.txt. They just use setup.py instead.

whit537 commented:
Another example: https://github.com/pypa/warehouse/blob/master/requirements/main.txt

mattbk commented:
Trying to catch up a little here. Is part of Gratipay project signup going to be associating your project with the results of gdr.rocks, or will it be a manual "Here's a list of projects you depend on (https://github.com/gratipay/gratipay.com/pull/4135#issuecomment-254403236), now go search for them on Gratipay?"

whit537 commented:
> This page describes PyPI's OAuth 1.0a facility. To use this facility you must first contact the PyPI maintainers (Richard or Martin) to obtain a consumer key and secret pair.

https://wiki.python.org/moin/PyPIOAuth

Can we use this to link packages to Gratipay accounts?

whit537 commented:
> Is part of Gratipay project signup going to be associating your project with the results of gdr.rocks, or will it be a manual "Here's a list of projects you depend on (https://github.com/gratipay/gratipay.com/pull/4135#issuecomment-254403236), now go search for them on Gratipay?"

*Project* sign-up will include the possibility of linking PyPI packages to your Gratipay account. We'll have pages like [`/on/pypi/aspen`](https://gratipay.com/on/twitter/beekme/) that can be linked to Teams (Projects), similar to the [`/on/twitter/beekme`](https://gratipay.com/on/twitter/beekme/) pages we have now, which can be linked to ~users.

That's not what https://github.com/gratipay/gratipay.com/pull/4135#issuecomment-254403236 is for, though. The "Open Source Inventory Report" is aimed at _companies_. The idea is that we'll give companies a way to set up a weekly payment that gets distributed _across the whole "bundle."_ Basically each week we'll:

1. recrawl the company's private repos,
1. find all of their dependency definition files (`setup.py`, `requirements.txt`, `package.json`, etc.),
1. feed those files to http://gdr.rocks/,
1. get back a comprehensive list of all of their dependencies,
1. see what dependencies are associated with a Gratipay Team (Project)
1. charge the company's credit card, and
1. split the money equally between all of the relevant Teams (Projects).

With this PR we're tackling (3) and (4). Once gdr.rocks is performant enough, I think that should be useful and interesting enough on its own to deploy. In terms of priorities after that, I'll probably go down before going back up. (5) is going to be tricky, since we need to figure out how to prove that someone owns a PyPI package. (6) and (7) will also take some work because we'll have to introduce a new concept of "bundles" or some such in between ~users paying and Teams (Projects) receiving. (1) and (2) will also take some work but I think that's the last to be done.

whit537 commented:
> I think that should be useful and interesting enough on its own to deploy.

Actually, we'll probably need to at least go through (5) for an MVP. Without at least a link over to a Team (Project) we have not funding loop that's even capable of being closed. Once we have at least that little loop, we can improve on it with the other pieces.

Changaco commented:
> (5) is going to be tricky, since we need to figure out how to prove that someone owns a PyPI package.

How about asking developers to add a link to the project's Gratipay account in the package metadata? That was the idea in <https://github.com/OpenCollective/OpenCollective/issues/178>.

whit537 commented:
> How about asking developers to add a link to the project's Gratipay account in the package metadata? 

That's a possibility, yes. It's the way [RubyGems](https://github.com/rubygems/rubygems.org/issues/500#issuecomment-65150164) went, as well as [Packagist](https://github.com/composer/packagist/issues/267#issuecomment-254454467). The alternative would seem to be looking for email addresses in the metadata and sending links there.

mattbk commented:
(5) is what I was asking about, thanks for the explanation at https://github.com/gratipay/gratipay.com/pull/4135#issuecomment-254563722.

mattbk commented:
> Actually, we'll probably need to at least go through (5) for an MVP.

True, but I don't think it needs to be automated for an MVP. The rate of return on doing it manually is probably pretty good.

whit537 commented:
> add a link to the project's Gratipay account in the package metadata

I'm not positive, but I'm pretty sure that [PyPI](https://docs.python.org/2/distutils/packageindex.html) doesn't allow for arbitrary package metadata. We'd have to override one of the existing [fields](https://docs.python.org/2/distutils/setupscript.html#additional-meta-data), hence the suggestion to use the emails that *are* in metadata.

whit537 commented:
> the suggestion

(On the phone under https://github.com/gratipay/gdr.rocks/issues/2#issuecomment-254577337.)

kaguillera commented:
So I think that I finally caught up and I may be jumping the gun and thinking to far ahead but I have a few questions that may or may not have been considered.

1. If we get a comprehensive list of all of their dependencies (1), (2), (3), (4) and (5) what about those open source softwares that are not part of projects (i.e. are not in the requirements.txt, setup.py, etc)  that are used by the companies. e.g. VIM?
2. Will we allow the companies to add/fund OSS other than the ones that they use?
3. Will we allow the companies to change how funds are distributed to different OSS projects if they think that they depend on some more than others?

whit537 commented:
(1) [Yes](https://github.com/gratipay/gratipay.com/pull/4135#issuecomment-254391020). Let's keep that in mind as we decide on pricing expectations.

(2) Yes! This is overlayed on Gratipay's normal Team (Project) system, where you can fund projects directly.

(3) Perhaps, though I doubt soon. (2) would be a workaround.

whit537 commented:
If this takes off it could exert pressure towards lots of little packages.

whit537 commented:
What about https://libraries.io/? Can we build on that somehow?

whit537 commented:
Okay! Yeah, looks like they [haven't](https://github.com/librariesio/support/issues/56) done the [necessary](https://github.com/gratipay/gdr.rocks/issues/2) for Python, but they have an [API](https://libraries.io/api). Let's see if we can use that for some combination of [(3) (4)](https://github.com/gratipay/gratipay.com/pull/4135#issuecomment-254563722) ...

whit537 commented:
Reached out: https://twitter.com/whit537/status/788809108170248192.

whit537 commented:
Filed bug: https://github.com/librariesio/libraries.io/issues/838.

whit537 commented:
Um, yeah. 😳 

> You can now browse the resolved transitive dependency tree for any version of any library right on the site: https://libraries.io/npm/base62/tree?kind=Development

![cug6ncwxeaaqzex](https://cloud.githubusercontent.com/assets/134455/19532468/dc273a6c-9609-11e6-95e8-a35516da4ec5.jpg)

https://twitter.com/librariesio/status/784113235188482048

whit537 commented:
> All API requests must include `api_key` parameter, get your api key from your [account page](https://libraries.io/account)

The account page link dies as well. Let's test this statement ...

whit537 commented:
Yessssssssss!

https://libraries.io/api/pypi/aspen is 200💃 

```json
{
    "name":"aspen",
    "platform":"Pypi",
    "description":"A filesystem router for Python web frameworks",
    "homepage":"https://github.com/AspenWeb/aspen.py",
    "repository_url":"https://github.com/AspenWeb/aspen.py",
    "normalized_licenses":[
        "Other"
    ],
    "rank":6,
    "latest_release_published_at":"2016-09-09T19:02:22.000Z",
    "latest_release_number":"1.0rc2",
    "language":null,
    "status":null,
    "package_manager_url":"https://pypi.python.org/pypi/aspen/",
    "stars":0,
    "forks":0,
    "keywords":[

    ],
    "latest_stable_release":{
        "number":"1.0rc2",
        "published_at":"2016-09-09T19:02:22.000Z"
    },
    "versions":[…]
}
```

whit537 commented:
Login bug fixed!

whit537 commented:
[Scheduling](https://twitter.com/teabass/status/788813114477256705) a call ...

whit537 commented:
We're on for 2pm UK time tomorrow!

whit537 commented:
There's an API for dependencies:

https://libraries.io/api/npm/grapesjs/0.3.21/dependencies

whit537 commented:
Adding an API for _resolved_ dependencies ticketed as https://github.com/librariesio/libraries.io/issues/839.

whit537 commented:
I think we should focus on JavaScript for our MVP. Like it or not, JavaScript—and not Python—is the Silicon Valley of open source.

"We didn't want someone else to be the Y Combinator of Silicon Valley. _We_ wanted to be the Y Combinator of Silicon Valley."

![screen shot 2016-10-19 at 9 46 06 pm](https://cloud.githubusercontent.com/assets/134455/19543447/7b987634-9645-11e6-82ee-27c0929d949d.png)

whit537 commented:
Besides that, the future looks a lot more like Libraries.io (two years old; their core product) rather than gdr.rocks (two days old; not our core product). Libraries.io [does not](https://github.com/librariesio/support/issues/56) support Python dependencies. That's the problem I've been working on in https://github.com/gratipay/gdr.rocks/issues/2, but I think it makes more sense to drop that work for this week, and talk to Libraries.io about donating it to them. We should be able to use the Libraries.io API today for `/on/npm/*` pages. It's true that they [don't have](https://github.com/librariesio/libraries.io/issues/839) resolved dependencies in the API yet, but even if they can't get that done in time for us to use next week, we can probably scrape the HTML.

whit537 commented:
@rohitpaulk @aandis Yinz have any spare cycles this weekend, either to review/merge or pick up some of the [tasks](https://github.com/gratipay/gratipay.com/pull/4135#issuecomment-254563722) here? We've got a couple big opportunities with [All Things Open] next week (https://github.com/gratipay/inside.gratipay.com/issues/757) and [Y Combinator](https://github.com/gratipay/inside.gratipay.com/issues/836) soon after. The further along we have this company giving experience, I think the better traction we're going to get in both contexts. Good be a turning point for us!

whit537 commented:
29c9e6f7ec492c0b76b3f46d588596789f3c3613 😍 💃 

![screen shot 2016-10-19 at 10 43 56 pm](https://cloud.githubusercontent.com/assets/134455/19544514/8a5839ea-964d-11e6-9421-316608131ee4.png)

whit537 commented:
> Libraries.io does not support Python dependencies. 

In fairness, this is noted on their [compatibility matrix](https://libraries.io/about).

whit537 commented:
EEEEEEEEEEEEEEEEEEEEEEEEE

![screen shot 2016-10-19 at 11 18 58 pm](https://cloud.githubusercontent.com/assets/134455/19545207/6f8f0620-9652-11e6-9e5a-5cfba0eaf418.png)

whit537 commented:
"Go" should post this to an endpoint under `/~username/giving/bundles/`, i.e., feed you back into that UI (which doesn't exist yet). A certain type of `bundle` can have a package.json as its definition.

whit537 commented:
fbc8c68525637783a62fc09538f66ed85dbec609

![screen shot 2016-10-19 at 11 29 34 pm](https://cloud.githubusercontent.com/assets/134455/19545408/ed505248-9653-11e6-81a0-0623006d5c81.png)

whit537 commented:
To the end of each dependency list, we should append: [the package manager, the language itself, Libraries.io, and Gratipay].

> what about Vim and Emacs?
> e.g. VIM?

I note that editors have plugin repos on Libraries.io, so that if we're appending to a dependency list these would be covered. Probably the target customer there, though, is the individual developer. We could/should cultivate a shared norm within open source that every dev pays a certain amount per week into their preferred editor ecosystem. With the Apple comments at https://github.com/gratipay/inside.gratipay.com/issues/844#issuecomment-253926018 in mind, how do we price this? Maybe 99¢ per week? That'd be ~$50 per year, which is cheaper than the list price of Sublime ([$70](https://www.sublimetext.com/buy?v=3)) and Ultraedit ([$99](http://www.ultraedit.com/catalog/)), though you can use each of those for more than a year. That'd also be $200,000/wk if we reach 0.1% of GitHub users; $200,000/wk if we reach 1%; and $2 M/wk if we reach 10%. We don't have the same nice `package.json` hook in a GitHub repo like we do for companies. Updating a bundle would probably be more manual for this case.

That still leaves operating systems and their ecosystems, which again have different characteristics between individual users (Windows 10 and macOS are [free](https://www.wired.com/2013/10/apple-ends-paid-oses/)) and companies (for whom Windows 10 is [$7/mo](http://www.forbes.com/sites/gordonkelly/2016/07/14/microsoft-confirms-windows-10-new-monthly-charge/)).

whit537 commented:
> To the end of each dependency list, we should append: [the package manager, the language itself, Libraries.io, and Gratipay].

Should we stabilize that as a percentage, maybe 1 or 2% each? Otherwise we have high variability across bundles. npm famously has lots of little packages, which means The Four(!) would see a smaller percentage from `package.json` files than from `requirements.txt` or `.gemspec` files. On the other hand, many of those npm packages will have been written by the same author, so in aggregate I expect the difference compared to individual developer payouts under more monolithic ecosystems won't be that great.

How about 1% each for the package manager and the language, and 2% each for Libraries.io and Gratipay? Including credit card fees of ~3% that would come out to ~91% going into the dependencies themselves. I'm thinking that at least for PyPI a lot of projects will depend on PyPA packages such as `setuptools`, `pip`, and `virtualenv`, so they would actually see more than 1%.

I'm also thinking of the 5% fee that one sees from Kickstarter, Patreon, etc. Compared to that, 2% to Gratipay itself is low.

What's the right way to price this?

whit537 commented:
We could also set the 1% or 2% as minimums. For something with only 50 dependencies we'd do an even split 54 ways.

rohitpaulk commented:
> @rohitpaulk @aandis Yinz have any spare cycles this weekend, either to review/merge or pick up some of the tasks here?

I'll definitely be available for reviews - Not sure about picking up tasks, I'll know by saturday morn :)

whit537 commented:
> use the emails that _are_ in metadata

Eureka! We don't even need to send special verification emails, because we _already_ send verification emails. We can just match already-verified emails against what we can pull out of metadata. So for example, when one of the Express maintainers or author is signed into Gratipay and views `/on/npm/express`, they'll see an extra button "You own this! Start accepting payments? [ Yes! ]".

whit537 commented:
Okay! Just wrapped up a call with Andrew Nesbitt and Ben Nickolls from Libraries.io. It was great! The three big things they're working on are:

1. **discovery**—[Libraries.io](http://libraries.io)
1. **maintainability**—[Dependency CI](https://dependencyci.com/)
1. **sustainability**—further off and less well-defined

Sustainability also includes gifts-in-kind ([e.g.](http://24pullrequests.com/)), but of course money is central to that discussion. It seems like there could be a fruitful partnership here, with them specializing in tracking the ecosystem and us specializing in moving money.

In terms of next week, https://github.com/librariesio/libraries.io/issues/14 prevents us from using Libraries.io for the `/on/*pm/foo/` pages, so I think we should limit ourselves to just `/on/npm/foo/` and use the npm API directly. We can still use Libraries.io for search ("Powered by") on `/on/npm/`. Once https://github.com/librariesio/libraries.io/issues/14 is done we can expand to other package managers. 

The idea of companies setting up a single payment to a whole bundle is exciting, and also further off.

> We don't even need to send special verification emails

Also, we will in fact need to send an additional verification email since package managers don't generally verify the email addresses that are added to metadata.

whit537 commented:
> In terms of next week,

(https://github.com/gratipay/inside.gratipay.com/issues/852#issuecomment-255098337 is the missing context.)

aandis commented:
@whit537 I should be able to get something done. What's our goal for this weekend?

whit537 commented:
@aandis Awesome! Our goal is to launch https://github.com/gratipay/gratipay.com/pull/4147 before next Thursday (week from today). Let's pick up over there?

whit537 commented:
@aandis Actually, scratch that. This'll require multiple PRs, so I made a coordinating ticket at https://github.com/gratipay/gratipay.com/issues/4148.

techtonik commented:
Wow! Time to get back to Gratipay now that https://github.com/gratipay/inside.gratipay.com/issues/317 became actionable. =)

`-` "Gratipay is a great way to fund open source"
`+` "Funding [your invisible infrastructure](https://medium.com/@nayafia/how-i-stumbled-upon-the-internet-s-biggest-blind-spot-b9aa23618c58#.ttaamairl)"

whit537 commented:
Perhaps we should take the angle that this is a matter of [corporate social responsibility](https://en.wikipedia.org/wiki/Corporate_social_responsibility).

whit537 commented:
I mean, right?

> **Corporate social responsibility** (**CSR**, also called **corporate conscience**, **corporate citizenship** or **responsible business**) is a form of corporate self-regulation integrated into a business model. CSR policy functions as a self-regulatory mechanism whereby a business monitors and ensures its active compliance with the spirit of the law, ethical standards and national or international norms. With some models, a firm's implementation of CSR goes beyond compliance and engages in "actions that appear to further some social good, beyond the interests of the firm and that which is required by law." The aim is to increase long-term profits and shareholder trust through positive public relations and high ethical standards to reduce business and legal risk by taking responsibility for corporate actions. CSR strategies encourage the company to make a positive impact on the environment and stakeholders including consumers, employees, investors, communities, and others.

https://en.wikipedia.org/wiki/Corporate_social_responsibility

whit537 commented:
We've received some awesome feedback from @arasmussen ("[this guy](https://github.com/gratipay/inside.gratipay.com/issues/867#issuecomment-261369390)" ;) in private email. Sharing with permission ...

> Hey Chad,
>
>I'd gladly contribute to a project on Gratipay, but I'd like to contribute to a project that we actually use and benefit from at Loopify.
>
>Specifically, here's a list of npm modules I'm grateful for:
>- async
>- babel-* (also css-loader, file-loader, saas-loader, style-loader, url-loader, etc)
>- iframe-resizer
>- mongoose
>- nodemon
>- react
>- react-helmet
>- react-modal
>- react-redux
>- react-router
>- react-router-redux
>- redux
>- redux-thunk
>- webpack
>
>I'd suggest you hit up all these project maintainers and say "I've got a startup founder who wants to pay for your project if you sign up for Gratipay". Then let me know! :)

--------------------

>>Gosh—thank you! 🙇
>>
>>This actually might fit in super well with the next feature we're turning our attention to, which is [integration with npm](https://github.com/gratipay/gratipay.com/issues/4148). The plan is to let you post a package.json, pay one amount, and have it distributed among all of the packages you depend on. Does that sound like a good way to address the "actually use and benefit" criterion?

--------------------

>>Like ... how did you come up with that list? Is it the same as or different than your package.json? Is it enough to start with package.json, or do you want to hand-curate the list to give to?

--------------------

>Kinda. It'd be nice if I could remove some projects from the list though. Some projects are extremely simple and don't really get maintained, whereas other projects are a full time effort from teams of people. The latter is the type I'd rather support.
>
>There are also npm modules in my package.json that I've simply forgotten to "npm uninstall". So I think being able to upload the package.json is cool, but then let me select which projects from that list I'd like to support.

--------------------

>I started with my package.json and trimmed a significant number of projects that are either very simple, unmaintained, or we don't actually use actively. See my other email :P

--------------------

>>Awesome. Our targets are:
>>
>>-1. load up npm in the Gratipay db—done!
>>0. mirror all npm packages on Gratipay 
>>1. enable npm packages to opt-in to receive funding
>>2. enable pledging to projects that haven't opted in yet
>>3. aggregate opt-in so maintainers can opt-in many at once
>>4. maybe naive package.json parsing with no deep traversal and equal split
>>5. possibly deep traversal (deps of deps)
>>6. eventually maybe fine-tuning of "bundles" like you're talking about
>>
>>Honestly, you should be able to manually set up payments as you want after (1) or maybe (2).
>>
>>May I post your comments to public GitHub for the rest of the team to see? Feedback is oxygen!

-----------------------

> Sure!

kaguillera commented:
This is great. Lets do it.

nobodxbodon commented:
- if package owner doesn't come to get their piece of donation, will it expire?
- will the total fund be distributed equally among all the projects in the whole dependency tree?
- any reason to paste package.json instead of uploading the file?

whit537 commented:
> - if package owner doesn't come to get their piece of donation, will it expire?

Hmm ... that's a new idea, I think. We did have "pledging" in the past: a promise to give, but we don't actually move money until the receiver opts in. Pledges didn't expire, but receivers could "lock" their `/on/foo/bar/` page to opt-out even from pledges.

> - will the total fund be distributed equally among all the projects in the whole dependency tree?

For the manual setup we're talking about at https://github.com/gratipay/gratipay.com/pull/4135#issuecomment-262672635, amounts would be manually set by the giver. More complicated distributions fall under #1493.

> - any reason to paste package.json instead of uploading the file?

Easier to build. :-)

bbangert commented:
I understand the appeal of this concept to companies, but I'm not sure I understand how it's a great thing for developers that want real project sustenance vs. some beer money. 

An apt example might be musicians.... who get peanuts from the divided lump-sum a listener throws at the streaming service. The musicians always *prefer* direct money to them, with concert tickets and album sales second, and streaming music a distant reluctant thing. Given the size of these dependency chains for some projects, it could also be as useless to many projects as most class action settlements are to the actual plaintiffs. A good/huge chunk of money split so many ways its not enough to make any real difference to the people that actually get it.

Ignoring that, the next biggest issue I'd imagine is what @nobodxbodon points out, how are the funds to be split amongst the dependency tree? By LoC? How does one blindly say what is the most significant, took the most time, or requires the most additional resources (actual new funding)?

Is the person at the company responsible for allocating money going to be uploading requirements files for all their project code-bases? How exactly does that work? Can companies instead choose existing 'stacks' they know they use, and trust that someone managing the funds for that 'stack' is overseeing fair allotment of funds?

Is this feature going to be appealing enough to OSS developers that they encourage companies to donate via this route instead of direct donations just to their project? (Where they'd get the entirety of it without having to worry about how its divided amongst a dependency chain)

I'm quite curious about what this feature entails since #637 is requesting a specific amount of money to implement what seems like a very vaguely defined feature with unknown desirability amongst the OSS developers it's supposed to benefit.

kaguillera commented:
@bbangert IMHO.

> I understand the appeal of this concept to companies, but I'm not sure I understand how it's a great thing for developers that want real project sustenance vs. some beer money.

The distant/difference between zero and one is infinite (at least in mathematics). In some case those developer are not getting any contributions for companies that want to contribute to them so something is better than nothing. That in itself is a step in the right direction. 

> Is this feature going to be appealing enough to OSS developers that they encourage companies to donate via this route instead of direct donations just to their project? (Where they'd get the entirety of it without having to worry about how its divided amongst a dependency chain)

The problem with the OSS developers trying to encourage companies to donate is if the company actually does pay attention to the developers and want to donate how easy/difficult is it in the current infrastructure. The financial reporting for just that one project would be a major turn off for the company . This is what we are trying to do, make the who process easier for both OSS developer and company willing to donate.

That being said the rest of you questions are what we are trying to work out.

whit537 commented:
@bbangert It's interesting that your feedback from a developer's perspective actually aligns with @arasmussen's [feedback](https://github.com/gratipay/gratipay.com/pull/4135#issuecomment-262672635) from the company giver point of view:

> I started with my package.json and trimmed a significant number of projects that are either very simple, unmaintained, or we don't actually use actively.

> Some projects are extremely simple and don't really get maintained, whereas other projects are a full time effort from teams of people. The latter is the type I'd rather support.
>
>There are also npm modules in my package.json that I've simply forgotten to "npm uninstall". So I think being able to upload the package.json is cool, but then let me select which projects from that list I'd like to support.

In other words, "enlightened" companies *want* to direct their payments to the right place. They want their money to have the most impact. I think there's still a roll for surfacing requirements via the package repos, because [recognition is easier than recall](https://www.nngroup.com/articles/recognition-and-recall/). If all you have to go on is the packages you _remember_ that you use, then your giving will be (even more) biased towards the projects with the best marketing.

> Is this feature going to be appealing enough to OSS developers that they encourage companies to donate via this route instead of direct donations just to their project? (Where they'd get the entirety of it without having to worry about how its divided amongst a dependency chain)

I think, done right, this can benefit smaller projects, where there is no time or budget to do the sales necessary to get companies to give just to them. An alternative would be to band together under a shared entity (PSF?), run marketing and sales through there, and divvy up the money however is agreed upon by that group.

whit537 commented:
I just got off the horn with @jdorfman, who pitched me an idea about providing assurance to companies that the long-tail of open source is sustainable for the long term, and I was able to share with him what we're working on here in terms of building on package managers towards that very end. :-)

Helpful in the back-and-forth was putting a point on the fact that what we need to balance here is:

1. making it easy for companies to get money into the ecosystem without the "chore" (@jdorfman's word) of sifting through the long tail to figure out where the money should go, and
1. allocating money fairly, to balance core tech with hot new projects, as @bbangert especially has been advocating here and even moreso at https://github.com/gratipay/inside.gratipay.com/issues/637#issuecomment-263469588.

jdorfman commented:
No one likes chores.

mattbk commented:
> No one likes chores.

Part of avoiding chores seems like setting up company [annual](https://github.com/gratipay/gratipay.com/issues/2401) or [monthly](https://github.com/gratipay/gratipay.com/issues/1199) invoicing. That seems important for the corporate side of things.

mattbk commented:
+1 from https://twitter.com/bbatsov/status/819392359087517696.

whit537 commented:
This is proving to be a larger task than it makes sense to have a single PR for. Closing in favor of our first epic: https://github.com/gratipay/inside.gratipay.com/issues/987.

whit537 commented:
Had a strategy session w/ @nobodxbodon last week the day after [Sustain](https://github.com/gratipay/inside.gratipay.com/issues/920). We met at [Coupa](https://www.coupacafe.com/locations/palo-alto/)—same place as [seven months ago](https://github.com/gratipay/inside.gratipay.com/issues/878#issuecomment-259299542). :o)

Caught up on the past six months and talked about priorities. What we converged on is as follows:

- Day 0 is when we complete the full "Give to package.json" loop for the first time.
  - corporate giver discovers packages via package.json
  - sets up payment to packages*
  - package owners are notified
  - package owners accept or reject (lock) the payment
- Until Day 0, our top priority is to get to Day 0.
  - Build out the product!
  - Warm up some companies to be our first customers.
    - Go through a sales exercise with smaller tech companies to complement the sales exercise we did with large companies.
    - Notify large company contacts when we hit Day 0 but don't count on interest--our product for them is [Open Source Perks](https://github.com/gratipay/gratipay.com/issues/1153), and this is just a step on the way.
  - Do not pay down tech debt yet.

\* *Our current payment flow is idiosyncratic and therefore frictive. dmk246 suggests making it more like a normal "shopping cart" checkout flow, which would be awesome but should maybe (probably) wait until after Day 0 ... except that one-time payments might be on the critical path for this to reduce friction for company givers.*

*This represents some significant shifts to our current flow that will take some significant effort to accomplish. Risk factor: this could spiral out of control into a full redesign.*
